engine/googletest_list.cpp

```diff
@@ -0,0 +1,122 @@
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include "engine/googletest_list.hpp"
+
+#include <fstream>
+#include <regex>
+#include <string>
+#include <utility>
+
+#include "engine/exceptions.hpp"
+#include "model/metadata.hpp"
+#include "model/test_case.hpp"
+#include "utils/config/exceptions.hpp"
+#include "utils/format/macros.hpp"
+
+namespace config = utils::config;
+namespace fs = utils::fs;
+
+
+namespace {
+
+/// A regular expression that should match either a test suite or a test case.
+const std::string name_expr = "([[:alpha:][:digit:]_]+[[:alpha:][:digit:]_/]*)";
+
+/// The separator between a test suite and a test case.
+const std::string testsuite_testcase_separator = ".";
+
+/// A complete regular expression representing a line with a test suite
+/// definition.
+///
+/// e.g.,
+///   "TestSuite."
+/// or
+///   "TestSuite/Prefix."
+/// or
+///   "TestSuite/Prefix.    # TypeParam = .+"
+const std::string testsuite_expr =
+    name_expr + "\\.([[:space:]]+# TypeParam = .+)?";
+
+/// A complete regular expression representing a line with a test case
+/// definition.
+///
+/// e.g.,
+///   "  TestCase"
+/// or
+///   "  TestCase/0"
+/// or
+///   "  TestCase/0  # GetParam() = 4"
+const std::string testcase_expr =
+    "  " + name_expr + "([[:space:]]+# GetParam\\(\\) = .+)?";
+
+}  // anonymous namespace
+
+
+/// Parses the googletest list of test cases from an open stream.
+///
+/// \param input The stream to read from.
+///
+/// \return The collection of parsed test cases.
+///
+/// \throw format_error If there is any problem in the input data.
+model::test_cases_map
+engine::parse_googletest_list(std::istream& input)
+{
+    std::regex testcase_re(testcase_expr), testsuite_re(testsuite_expr);
+    std::smatch match;
+    std::string line;
+    std::string testcase_name, test_suite;
+
+    model::test_cases_map_builder test_cases_builder;
+    while (std::getline(input, line).good()) {
+        if (std::regex_match(line, match, testcase_re)) {
+            std::string test_case;
+
+            if (test_suite.empty()) {
+                throw format_error("Invalid testcase definition: not preceded "
+                                   "by a test suite definition");
+            }
+            test_case = std::string(match[1]);
+            test_cases_builder.add(test_suite + test_case);
+        } else if (std::regex_match(line, match, testsuite_re)) {
+            test_suite = std::string(match[1]) +
+                         testsuite_testcase_separator;
+        }
+        // else, ignore the line; something might have used output a diagnostic
```

Comment:
nit: Encode as much as possible of your comment in actual code. That means defining an empty `else` branch and putting the comment within it, without the "else" part.